File: tests/typesafety/python/exercise_shipping_solution.py
# expect: passes
# title: Exercise 2, one solution
# improves: exercise_shipping_start
"""Units as types, options as named fields. The call now says what every
value means, and swapped units are a compile error."""

from dataclasses import dataclass
from typing import NewType

Kilograms = NewType("Kilograms", float)
Kilometers = NewType("Kilometers", float)


@dataclass(frozen=True)
class Handling:
    express: bool = False
    insured: bool = False
    fragile: bool = False


def shipping_cents(weight: Kilograms, distance: Kilometers, handling: Handling) -> int:
    rate = 3 if handling.express else 1
    surcharge = (25 if handling.insured else 0) + (40 if handling.fragile else 0)
    return int(weight * distance * rate) + surcharge


quote = shipping_cents(
    Kilograms(2.5), Kilometers(120.0), Handling(express=True, fragile=True)
)

File: tests/typesafety/typescript/exercise_shipping_solution.ts
// expect: passes

declare const kilogramsBrand: unique symbol;
declare const kilometersBrand: unique symbol;

type Kilograms = number & { readonly [kilogramsBrand]: true };
type Kilometers = number & { readonly [kilometersBrand]: true };

function kilograms(n: number): Kilograms {
  return n as Kilograms;
}

function kilometers(n: number): Kilometers {
  return n as Kilometers;
}

type Handling = {
  readonly express?: boolean;
  readonly insured?: boolean;
  readonly fragile?: boolean;
};

function shippingCents(weight: Kilograms, distance: Kilometers, handling: Handling): number {
  const rate = handling.express ? 3 : 1;
  const surcharge = (handling.insured ? 25 : 0) + (handling.fragile ? 40 : 0);
  return Math.trunc(weight * distance * rate) + surcharge;
}

export const quote = shippingCents(kilograms(2.5), kilometers(120.0), {
  express: true,
  fragile: true,
});
